### ISSUE TITLE
Support: optional numbers.

### ISSUE THREAD

**Issue by oscbyspro** [oscbyspro/DiffableTextViews#86]
When an optional value is nil the view should show a placeholder. While supporting `Optional<T>` is orthogonal to number processing, and arguably does little to improve UX, it’s like a 0th class citizen in Swift-ville and people just expect it to work, even if SwiftUI.TextField doesn’t support it.

**Comment by oscbyspro** [comment]
The easiest way to accomplish this is probably to create a new FormatStyle wrapper, then extend Optional to use it. It would have to modify the format and parse methods to consider optionality. Haven’t thought much about it yet, but I think it will be trivial to implement.

**Comment by oscbyspro** [comment]
If bounds.autocorrect(\_:) were moved to the end of the interpret(\_:) method, then I it would be enough to allow Number.init to return an optional value (with a static isOptional, like isUnsigned and isInteger). Haven’t checked if moving it is sensible/valid, however. It’s a bit more complicated otherwise, because optionals are not comparable.

**Comment by oscbyspro** [comment]
As of dccf38ddac3090cc7ed5ef739a43b0d0ac11c3a6 (dev) it should be sufficient to provide conditional behavior to the very first lines in `interpret(_:)` and `merge(_:)`, after `Number.init` has been made optional. It’s not trivial, but it’s also not complicated and should not increase the module's complexity all that much.

## Proposal (subject to change)

**interpret(\_:)**

Let the wrapped format provide an optional commit per value. The commit should be nil unless the value itself is optional and nil, at which point the commit should contain it and an empty snapshot. If the commit is not nil it should be returned, skipping the body of this method.

**merge(\_:)**

Optional values should return an empty commit if Number.init returns nil. Non-optional values should crash the application, because then Number.init should never return nil for non-optional values.

**Comment by oscbyspro** [comment]
...forgot I still have to decide on how to handle that optional numbers are not comparable. Given that it has to be comparable (or offer comparable functionality) it should be possible to delegate the proposed `interpret(_:)` specific part to Number.init a few lines below, reducing overall complexity. I probably should not use `Comparable` directly, considering that this part of the module is exposed as public.

**Comment by oscbyspro** [comment]
Another option is to wrap NumberTextStyle as OptionalNumberTextStyle, then give it a nice type alias. It might be easier given how orthogonal Optional is. Hm.

**Comment by oscbyspro** [comment]
Above solution works as expected. The text is interpreted as a nil value when there is no: integer digits, fraction separator, and no fraction digits. Otherwise it works exactly as the non-optional style. Signs are ignored because depending on format they may appear outside the user's input space, like "-$US 1,234,567.89" for example.

**Comment by oscbyspro** [comment]
It’s surprisingly uncomplicated and works well, but I need to make some adjustments to how invalid signs are handled on user input. As it stands, going from (nil) to (+1) in bounds (≤0) does not work. It works for non-optional numbers because the sign is always valid, which means that discarding inputs with invalid signs is fine. This is not true for optional numbers, however.

**Comment by oscbyspro** [comment]
Solution proposed in 97a97dbbab9b8e589ffa420cc239b6ade1aa7172 (dev) and will be added in v4.0.0.

**Comment by oscbyspro** [comment]
Solved type aliasing in bec3dbe3d8edd1ee412beb26447d191b1d227891. It is now possible to write

```swift
NumberTextStyle<T>
T.NumberTextStyle
```

```swift
NumberTextStyle<T?>
T?.NumberTextStyle.
```
